sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/util/DubboAttachmentsHelper.java

```diff
@@ -46,17 +54,22 @@ private DubboAttachmentsHelper() {
      *
      * @param invocation call information
      * @param isApache whether it is apache dubbo
+     * @param isProvider whether it is dubbo server
      * @return Map
      */
-    public static Map<String, String> resolveAttachments(Object invocation, boolean isApache) {
+    public static Map<String, String> resolveAttachments(Object invocation, boolean isApache, boolean isProvider) {
         if (invocation == null) {
             return Collections.emptyMap();
         }
         final Map<String, String> attachments = new HashMap<>();
         if (isApache) {
-            attachments.putAll(getAttachmentsFromContext(APACHE_RPC_CONTEXT));
+            if (isDubbo3(invocation.getClass())) {
+                setDubbo3Attachments(attachments, isProvider);
+            } else {
+                attachments.putAll(getAttachmentsFromContext(APACHE_RPC_CONTEXT, GET_CONTEXT_METHOD));
+            }
         } else {
-            attachments.putAll(getAttachmentsFromContext(ALIBABA_RPC_CONTEXT));
+            attachments.putAll(getAttachmentsFromContext(ALIBABA_RPC_CONTEXT, GET_CONTEXT_METHOD));
         }
         final Optional<Object> fieldValue = ReflectUtils.getFieldValue(invocation, ATTACHMENTS_FIELD);
         if (fieldValue.isPresent() && fieldValue.get() instanceof Map) {
```

Comment:
It is recommended that if nesting be reduced.

Reply:
fixed